File: VB/CardDAVServer.FileSystemStorage.AspNet/Logger.vb
Imports System
Imports System.Configuration
Imports System.IO
Imports System.Web
Imports ITHit.WebDAV.Server
Imports ITHit.WebDAV.Server.Logger

''' <summary>
''' Logger singleton.
''' We need this singleton because logging is used in various modules, like authentication etc.
''' </summary>
Module Logger

    ''' <summary>
    ''' Whether debug logging is enabled. In particular most request/response bodies will
    ''' be logged if debug logging is on.
    ''' </summary>
    Private ReadOnly debugLoggingEnabled As Boolean = "true".Equals(ConfigurationManager.AppSettings("DebugLoggingEnabled"),
                                                                   StringComparison.InvariantCultureIgnoreCase)

    ''' <summary>
    ''' Path where log files will be stored.
    ''' </summary>
    Private ReadOnly logPath As String = ConfigurationManager.AppSettings("LogPath")

    ''' <summary>
    ''' Synchronization object.
    ''' </summary>
    Private ReadOnly syncRoot As Object = New Object()

    ''' <summary>
    ''' Gets logger instace.
    ''' </summary>
    Public ReadOnly Property Instance As ILogger
        Get
            Dim context = HttpContext.Current
            Const LOGGER_KEY As String = "$DavLogger$"
            If context.Application(LOGGER_KEY) Is Nothing Then
                SyncLock syncRoot
                    If context.Application(LOGGER_KEY) Is Nothing Then
                        context.Application(LOGGER_KEY) = initLogger()
                    End If

                End SyncLock
            End If

            Return CType(context.Application(LOGGER_KEY), ILogger)
        End Get
    End Property

    ''' <summary>
    ''' Initializes logger.
    ''' </summary>
    ''' <returns>Instance of <see cref="ILogger"/> .</returns>
    Private Function initLogger() As ILogger
        Dim logger = New DefaultLoggerImpl()
        Dim context = HttpContext.Current
        If Not String.IsNullOrEmpty(logPath) Then
            logger.LogFile = Path.Combine(context.Server.MapPath(logPath), "WebDAVlog.txt")
        Else
            logger.LogFile = Path.Combine(context.Request.PhysicalApplicationPath, "WebDAVlog.txt")
        End If

        logger.IsDebugEnabled = debugLoggingEnabled
        Return logger
    End Function
End Module

File: CS/CalDAVServer.FileSystemStorage.AspNet/Logger.cs
using System;
using System.Configuration;
using System.IO;
using System.Web;

using ITHit.Server;
using ITHit.WebDAV.Server;
using ITHit.WebDAV.Server.Logger;

namespace CalDAVServer.FileSystemStorage.AspNet
{
    /// <summary>
    /// Logger singleton.
    /// We need this singleton because logging is used in various modules, like authentication etc.
    /// </summary>
    public static class Logger
    {
        /// <summary>
        /// Whether debug logging is enabled. In particular most request/response bodies will
        /// be logged if debug logging is on.
        /// </summary>
        private static readonly bool debugLoggingEnabled =
            "true".Equals(
                ConfigurationManager.AppSettings["DebugLoggingEnabled"],
                StringComparison.InvariantCultureIgnoreCase);

        /// <summary>
        /// Path where log files will be stored.
        /// </summary>
        private static readonly string logPath = ConfigurationManager.AppSettings["LogPath"];

        /// <summary>
        /// Synchronization object.
        /// </summary>
        private static readonly object syncRoot = new object();

        /// <summary>
        /// Gets logger instace.
        /// </summary>
        public static ILogger Instance
        {
            get
            {
                var context = HttpContext.Current;
                const string LOGGER_KEY = "$DavLogger$";
                if (context.Application[LOGGER_KEY] == null)
                {
                    lock (syncRoot)
                    {
                        if (context.Application[LOGGER_KEY] == null)
                        {
                            context.Application[LOGGER_KEY] = initLogger();
                        }
                    }
                }

                return (ILogger)context.Application[LOGGER_KEY];
            }
        }

        /// <summary>
        /// Initializes logger.
        /// </summary>
        /// <returns>Instance of <see cref="ILogger"/>.</returns>
        private static ILogger initLogger()
        {
            var logger = new DefaultLoggerImpl();
            var context = HttpContext.Current;

            if (!string.IsNullOrEmpty(logPath))
            {
                logger.LogFile = Path.Combine(context.Server.MapPath(logPath), "WebDAVlog.txt");
            }
            else
            {
                logger.LogFile = Path.Combine(context.Request.PhysicalApplicationPath, "WebDAVlog.txt");
            }

            logger.IsDebugEnabled = debugLoggingEnabled;

            return logger;
        }
    }
}
